IpmiFeaturePkg/Library/IpmiTransportLibMock/MockIpmi.c

```diff
@@ -0,0 +1,181 @@
+  )
+{
+  IPMI_SELF_TEST_RESULT_RESPONSE  *SelfTest;
+
+  ASSERT (*ResponseSize >= sizeof (IPMI_SELF_TEST_RESULT_RESPONSE));
+
+  SelfTest                 = Response;
+  SelfTest->CompletionCode = IPMI_COMP_CODE_NORMAL;
+  SelfTest->Result         = IPMI_APP_SELFTEST_NOT_IMPLEMENTED;
+  SelfTest->Param          = 0;
+
+  *ResponseSize = sizeof (IPMI_SELF_TEST_RESULT_RESPONSE);
+}
+
+/**
+  Mocks the result of IPMI_APP_GET_DEVICE_ID.
+
+  @param[in]       Data           The IPMI request data.
+  @param[in]       DataSize       The size of the IPMI request data.
+  @param[out]      Response       The response data buffer.
+  @param[in, out]  ResponseSize   On input, the available size of buffer.
+                                  On output, the size of written data in the buffer.
+**/
+VOID
+MockIpmiGetDeviceId (
+  IN VOID       *Data,
+  IN UINT8      DataSize,
+  OUT VOID      *Response,
+  IN OUT UINT8  *ResponseSize
+  )
+{
+  IPMI_GET_DEVICE_ID_RESPONSE  *DeviceId;
+
+  ASSERT (*ResponseSize >= sizeof (IPMI_GET_DEVICE_ID_RESPONSE));
+
+  DeviceId                       = Response;
+  DeviceId->CompletionCode       = IPMI_COMP_CODE_NORMAL;
+  DeviceId->DeviceId             = 0xAB;
+  DeviceId->DeviceRevision.Uint8 = 0;
+  DeviceId->FirmwareRev1.Uint8   = 0x0;
+  DeviceId->MinorFirmwareRev     = 0;
+  DeviceId->SpecificationVersion = 2;
+  DeviceId->DeviceSupport.Uint8  = 0;
+  DeviceId->ManufacturerId[0]    = 0xB;
+  DeviceId->ManufacturerId[1]    = 0xA;
+  DeviceId->ManufacturerId[2]    = 0xD;
+  DeviceId->ProductId            = 1;
+  DeviceId->AuxFirmwareRevInfo   = 0;
+
+  *ResponseSize = sizeof (IPMI_GET_DEVICE_ID_RESPONSE);
+}
+
+//
+// Registered handlers.
+//
+
+MOCK_IPMI_HANDLER_ENTRY  MockHandlers[] = {
+  { IPMI_NETFN_APP, IPMI_APP_GET_DEVICE_ID,        MockIpmiGetDeviceId },
+  { IPMI_NETFN_APP, IPMI_APP_GET_SELFTEST_RESULTS, MockIpmiGetSelfTest },
+};
+
+//
+// Generic routines for handling top level IPMI commands and responses.
+//
+
+/**
+  Processes a IPMI request and prepared the next response.
+
+  @param[in]  Command       The IPMI request to mock.
+  @param[in]  Size          The size of the IPMI request.
+
+  @retval   EFI_SUCCESS     Always
+**/
+EFI_STATUS
+MockIpmiCommand (
+  IN IPMI_COMMAND  *Command,
+  IN UINT8         Size
+  )
+{
+  UINT32  Index;
+  UINT8   ResponseDataSize;
+
+  ASSERT (Command->Lun == 0);
+
+  //
+  // Generic response handling. Add the 0 bit value to indicate this is a
+  // response.
+  //
+
+  gResponse.NetFunction = Command->NetFunction | 0x1;
+  gResponse.Command     = Command->Command;
+  gResponseSize         = IPMI_RESPONSE_HEADER_SIZE;
+
+  //
+  // Search for a specific handler.
+  //
+
+  for (Index = 0; Index < (sizeof (MockHandlers) / sizeof (MockHandlers[0])); Index++) {
+    if ((MockHandlers[Index].NetFunction == Command->NetFunction) &&
+        (MockHandlers[Index].Command == Command->Command))
+    {
+      ResponseDataSize = sizeof (gResponse) - 2;

```

Comment:
Is this "2" the IPMI_COMMAND_HEADER_SIZE? Might want to use that?

Reply:
its actually a little different. The structures used actually overlap with the response buffer so its size - 1. I'll try to clean the structures up a bit to avoid this.